crud/src/main/java/com/redhat/lightblue/crud/AbstractBulkJsonObject.java

```diff
@@ -0,0 +1,158 @@
+ */
+package com.redhat.lightblue.crud;
+
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Collections;
+import java.util.Comparator;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+
+import com.redhat.lightblue.util.JsonObject;
+import com.redhat.lightblue.Request;
+
+/**
+ * Base class for bulk request and responses. This class contains the
+ * code common to both, as bulk request and response are structurally
+ * similar. It deals with a JSON of the form:
+ *
+ *
+ * <pre>
+ *   {
+ *     "<entries>": [
+ *         {
+ *             "seq":0,
+ *             "op": "FIND",
+ *             "<entry>": { item }
+ *         }
+ *     ]
+ *   }
+ * </pre> 
+ */
+abstract class AbstractBulkJsonObject<T extends JsonObject> extends JsonObject {
+
+    private final List<T> entries=new ArrayList<T>();
+
+    /**
+     * Returns all entries in the bulk object
+     */
+    public List<T> getEntries() {
+        return entries;
+    }
+
+    /**
+     * Sets the entries in the bulk object. Copies the collection
+     */
+    public void setEntries(List<T> x) {
+        entries.clear();
+        entries.addAll(x);
+    }
+
+    public void add(T x) {
+        entries.add(x);
+    }
+
+    /**
+     * Overriding this method allows the concrete class to add any necessary items to the entry node.
+     *
+     * @param node The object node, an element of the entries array
+     * @param entry The entry
+     */
+    protected abstract void toJsonEntryNode(ObjectNode node,T entry);
+
+    /**
+     * Returns a JSON representation of this
+     */
+    public JsonNode toJson(String entriesName, String entryName) {
+        JsonNodeFactory factory=getFactory();
+        ObjectNode node = factory.objectNode();
+        ArrayNode arr=factory.arrayNode();
+        int seq=0;
+        for(T x:entries) {
+            ObjectNode entryNode=factory.objectNode();
+            entryNode.set("seq",factory.numberNode(seq++));
+            toJsonEntryNode(entryNode,x);
+            entryNode.set(entryName,x.toJson());
+            arr.add(entryNode);
+        }
+        node.put(entriesName, arr);
+        return node;
+    }
+
+    private static class SReq {
+        private final int seq;
+        private final JsonObject entry;
+
+        public SReq(int seq,JsonObject entry) {
+            this.seq=seq;
+            this.entry=entry;
+        }
+    }
+
+    /**
+     * Parses the actual entry from the given node
+     *
+     * @param entry The entry object, an element of the entries array
+     */
+    protected abstract T parseEntry(ObjectNode entry);
+    
+    /**
+     * Parses the bulk object from the given array node
+     */
+    protected void parse(ArrayNode entriesArray) {
```

Comment:
I left it there in case we need to implement parse for BulkResponse.
Currently, BulkRequest is only used as an input, and BulkResponse is only
used as an output. If this changes going forward, both may end up having
parse/toJson.

On Thu, Aug 27, 2015 at 11:49 AM, Dennis Crissman <notifications@github.com>
wrote:

> In
> crud/src/main/java/com/redhat/lightblue/crud/AbstractBulkJsonObject.java
> <https://github.com/lightblue-platform/lightblue-core/pull/437#discussion_r38124595>
> :
>
> > +            this.seq=seq;
> > +            this.entry=entry;
> > +        }
> > +    }
> > +
> > +    /**
> > +     * Parses the actual entry from the given node
> > +     *
> > +     * @param entry The entry object, an element of the entries array
> > +     */
> > +    protected abstract T parseEntry(ObjectNode entry);
> > +
> > +    /**
> > +     * Parses the bulk object from the given array node
> > +     */
> > +    protected void parse(ArrayNode entriesArray) {
>
> I see only two implementations of AbstractBulkJsonObject, and only 1 uses
> this functionality while the other throws a UnsupportedOperationException.
> Would it make more sense to just move this functionality directly to
> BulkRequest and not even provide it in BulkResponse? Or do you plan to have
> more implementations?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/437/files#r38124595>
> .
>